Sources/OpenAPIRuntime/Conversion/Converter+Common.swift

```diff
@@ -17,31 +17,53 @@ extension Converter {
 
     // MARK: Miscs
 
-    /// Validates that the Content-Type header field (if present)
-    /// is compatible with the provided content-type substring.
+    /// Returns the MIME type from the content-type header, if present.
+    /// - Parameter headerFields: The header fields to inspect for the content
+    /// type header.
+    /// - Returns: The content type value, or nil if not found or invalid.
+    public func extractContentTypeIfPresent(in headerFields: [HeaderField]) -> MIMEType? {
+        guard let rawValue = headerFields.firstValue(name: "content-type") else {
+            return nil
+        }
+        return MIMEType(rawValue)
+    }
+
+    /// Checks whether a concrete content type matches an expected content type.
     ///
-    /// Succeeds if no Content-Type header is found in the response headers.
+    /// The concrete content type can contain parameters, such as `charset`, but
+    /// they are ignored in the equality comparison.
     ///
+    /// The expected content type can contain wildcards, such as */* and text/*.
     /// - Parameters:
-    ///   - headerFields: Header fields to inspect for a content type.
-    ///   - substring: Expected content type.
-    /// - Throws: If the response's Content-Type value is not compatible with the provided substring.
-    public func validateContentTypeIfPresent(
-        in headerFields: [HeaderField],
-        substring: String
-    ) throws {
-        guard
-            let contentType = try getOptionalHeaderFieldAsText(
-                in: headerFields,
-                name: "content-type",
-                as: String.self
-            )
-        else {
-            return
+    ///   - received: The concrete content type to validate against the other.
+    ///   - expected: The expected content type, can contain wildcards.
+    /// - Returns: A Boolean value representing whether the concrete content
+    /// type matches the expected one.
+    public func isValidContentType(received: MIMEType?, expected: String) -> Bool {
+        guard let received else {
+            return false
         }
-        guard contentType.localizedCaseInsensitiveContains(substring) else {
-            throw RuntimeError.unexpectedContentTypeHeader(contentType)
+        let receivedContentType = received
+        guard let expectedContentType = MIMEType(expected) else {
+            return false
```

Comment:
> So this is going to return false if the expected input to the function is not a valid mimetype?

Correct.

> Can't we just have this function take a MIMEType for expected?

That would push this decision (initialize from string and handle it returning nil) from this one place to all of the call sites in the generated code, and we'd still have to decide what to do in that scenario, so it was easier to have it take a string and optionally convert here. Do you think we should fail if the expected content type isn't valid? That'd also be reasonable, and we could throw from here, so the callsites wouldn't need to change.

Reply:
If we think that it's an invalid use of the function to pass an invalid MIME type to `expected` then I think we should treat it differently from the validation logic, yes. I'd be happy with a thrown error in this case I guess.

Would this be a bug in the generator? If so, maybe we should be adding an assert too?

Reply:
Since this is an SPI function on `Converter`, which are all optimized to make the generated code as short/high signal as possible, I'd be happy to rename the function to reflect the fact that it expects to receive an unvalidated content type string, and be able to handle if it's invalid. Maybe I could rename the parameter label from `expected` to `expectedRaw` or something, and update the doc comments?

And sure, I can make it throw if the expected content type doesn't parse.

This could happen if the exact content type parsing logic in OpenAPIKit, and some more processing in the generator don't exactly match the runtime library's. That could easily happen since generator doesn't depend on an exact, but a floating version of runtime.

Ok, let me make these two changes (rename the label and make it throw on invalid).

Reply:
Ok fixed:
- runtime: https://github.com/apple/swift-openapi-runtime/pull/29/commits/e0dae8558525888f24b2d875375dfd0856b150c6
- generator: https://github.com/apple/swift-openapi-generator/pull/146/commits/248840c47b39f7d633c94b53a584f334e83776c4